vllm/model_executor/layers/quantization/gptq_bitblas.py

```diff
@@ -0,0 +1,527 @@
+            self.quant_config.TORCH_BITBLAS_STORAGE_DTYPE)
+        intweight = unpack_qweight(qweight,
+                                   self.quant_config.weight_bits).contiguous()
+        if self.bitblas_matmul.weight_transform is not None:
+            qweight = self.bitblas_matmul.weight_transform(
+                intweight.cpu()).cuda()
+        # scales in gptq old quant linear stored with
+        # (infeatures // group_size, outfeatures), should be transposed.
+        scales = scales.T.contiguous()
+        # qzeros should be de-quantized to int zeros.
+        intzeros = unpack_qzeros(qzeros,
+                                 self.quant_config.weight_bits).T.contiguous()
+        zeros: Optional[torch.Tensor] = None
+        if self.bitblas_matmul.config.zeros_mode == "original":
+            zeros = intzeros.to(torch.float16).contiguous()
+        elif self.bitblas_matmul.config.zeros_mode == "rescale":
+            assert zeros is not None, "zeros should not be None"
+            zeros[:, :] = intzeros.to(torch.float16)[:, :] * scales[:, :]
+        elif self.bitblas_matmul.config.zeros_mode == "quantized":
+            zeros = (torch.Tensor(
+                general_compress(
+                    intzeros.T.contiguous().cpu().numpy(),
+                    self.quant_config.weight_bits,
+                )).to(qweight.device).to(
+                    self.quant_config.TORCH_BITBLAS_STORAGE_DTYPE).contiguous(
+                    ))
+        else:
+            raise ValueError("Unsupported zeros type: {}".format(
+                self.bitblas_matmul.config.zeros_mode))
+
+        return qweight, scales, zeros
+
+    def apply(
+        self,
+        layer: torch.nn.Module,
+        x: torch.Tensor,
+        bias: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+
+        part_size_n = layer.output_size_per_partition
+        out_shape = x.shape[:-1] + (part_size_n, )
+
+        if layer.bitblas_state == GPTQBitBLASState.REPACK:
+            layer.bitblas_state = GPTQBitBLASState.READY
+
+            # Newly generated tensors need to replace existing tensors that are
+            # already registered as parameters by vLLM (and won't be freed)
+            def replace_tensor(name, new_t):
+                # It is important to use copy_() here since it ensures
+                # the same buffer is reused
+                getattr(layer, name).copy_(
+                    new_t.view(getattr(layer, name).dtype).view(
+                        getattr(layer, name).shape))
+                del new_t
+
+            # Repack weights
+            bitblas_qweight, bitblas_scales, bitblas_qzeros = (
+                self.repack_bitblas_from_gptq(
+                    layer.qweight,
+                    layer.scales,
+                    layer.qzeros,
+                ))
+            replace_tensor("qweight", bitblas_qweight)
+            replace_tensor("scales", bitblas_scales)
+            replace_tensor("qzeros", bitblas_qzeros)
```

Comment:
It would be best to move this into a `process_weights_after_loading` function we have specifically for this purpose, example in gptq_marlin.py https://github.com/vllm-project/vllm/blob/a84e598e2125960d3b4f716b78863f24ac562947/vllm/model_executor/layers/quantization/gptq_marlin.py#L287

Reply:
Thanks, I'll take a look. I'm currently working on the stream-k template in bitblas :)